core/toolbox/category.js

```diff
@@ -0,0 +1,269 @@
+   * @protected
+   */
+  this.rowDiv_ = null;
+
+  /**
+   * The html element for the toolbox icon.
+   * @type {?Element}
+   * @protected
+   */
+  this.iconSpan_ = null;
+
+  /**
+   * Container for any children categories.
+   * @type {?Element}
+   * @protected
+   */
+  this.subcategoriesDiv_ = null;
+
+  /**
+   * The config for all the category css classes.
+   * @type {!Blockly.ToolboxCategory.ClassConfig}
+   * @protected
+   */
+  this.classConfig_ = {
+    'container': 'blocklyToolboxCategory',
+    'row': 'blocklyTreeRow',
+    'icon': 'blocklyTreeIcon',
+    'label': 'blocklyTreeLabel',
+    'contents': 'blocklyToolboxContents',
+    'selected': 'blocklyTreeSelected',
+    'openIcon': 'blocklyTreeIconOpen',
+    'closedIcon': 'blocklyTreeIconClosed'
+  };
+
+  Blockly.utils.object.mixin(this.classConfig_, categoryDef['classConfig']);
+
+  /**
+   * Whether or not the category should display it's children.
+   * @type {boolean}
+   * @protected
+   */
+  this.expanded_ = categoryDef['expanded'] == 'true' || categoryDef['expanded'];
+
+  /**
+   * True if the category is visible, false otherwise.
+   * @type {boolean}
+   * @private
+   */
+  this.isVisible_ = true;
+
+  /**
+   * Parse the contents for this category.
+   * @type {string|
+   *        !Array<!Blockly.ToolboxItem>|
+   *        !Array<!Blockly.utils.toolbox.FlyoutItemDef>}
+   * @protected
+   */
+  this.contents_ = this.parseContents_(categoryDef, this.hasChildren_);
+};
+
+Blockly.utils.object.inherits(Blockly.ToolboxCategory,
+    Blockly.CollapsibleToolboxItem);
+
+/**
+ * @typedef {{
+ *            container:?string,
+ *            row:?string,
+ *            icon:?string,
+ *            label:?string,
+ *            contents:?string,
+ *            selected:?string,
+ *            openIcon:?string,
+ *            closedIcon:?string,
+ *          }}
+ */
+Blockly.ToolboxCategory.ClassConfig;
+
+/**
+ * The number of pixels to move the category over at each nested level.
+ * @type {number}
+ */
+Blockly.ToolboxCategory.NestedValue = 19;
+
+/**
+ * Parse the contents array depending on if the category has children, is a
+ * dynamic category, or if it's contents are meant to be shown in the flyout.
+ * @param {!Blockly.utils.toolbox.Category} categoryDef The information needed
+ *     to create a category.
+ * @param {boolean} hasChildren True if this category has subcategories, false
+ *     otherwise.
+ * @return {string|!Array<!Blockly.ToolboxItem>|
+ *          !Array<!Blockly.utils.toolbox.FlyoutItemDef>}
+ *     The contents for this category.
+ * @private
+ */
+Blockly.ToolboxCategory.prototype.parseContents_ = function(categoryDef,
+    hasChildren) {
+  var toolboxItems = [];
+  var contents = categoryDef['contents'];
+  if (hasChildren) {
+    for (var i = 0; i < contents.length; i++) {
+      var child = new Blockly.ToolboxCategory(contents[i], this.parentToolbox_, this);
+      toolboxItems.push(child);
+    }
+  } else if (categoryDef['custom']) {
```

Comment:
I don't see this type defined in the typedef when I check utlils/toolbox.js on category_init branch.

Reply:
added 'custom' field to the Blockly.utils.toolbox.Category typedef.

Reply:
What is custom? The other properties seem fairly self-explanatory, but I don't really have any idea of what it is.

Reply:
It is how we allow people to have dynamic content for a flyout. For example: 
`<category name="Variables" categorystyle="variable_category" custom="VARIABLE"></category>` 
I agree it isn't the best name, but for legacy reasons I think it probably needs to stay the way it is.